### DISCUSSION TITLE
want more examples/documentation about wave

### DISCUSSION THREAD

**Discussion by zillionare** [h2oai/wave#1383] [General]
I have tried plotly dash and then find h2o wave. h2o wave is great in these aspects:

1. asyncio interface and use Go as proxy server, this may bring much better performance than dash. It would be great if somebody could provide a benchmark though. Dash is based on flashk and all requests are handled in sync way.
2. better API designs. In dash we have to bound too many Input/Output deps on one callback, and multiple callback target on same control/property is not allowed, this cause the callback is not concise and sometime hard to write. Same that wave is better here, but I doesn't try too further.
3. More components/widgets.

however, Plotly dash is great with its forum and active community and it's documentation. I can got answers for almost every question easily. However, I found much less resource (stackoverflow, forum posts, etc) about wave. 

Currently it's great if you can provide document/example on:

1. how to arrange wave application fold structure, or how to modulize wave app?
2. how to write a multipage and routing to each other?
3. how to rewrite http headers, respond with http status?

by the way, this gist is great:

https://gist.github.com/mtanco/767538459821fcc51cc3d533fb584249

**Comment by lo5** [comment]
Thanks for your feedback @zillionare, and thanks for trying out Wave!

Wave is a relatively newer project compared to Dash, and we're improving the docs as we go. Hopefully as it gets more adoption, there'll be more content, blogs and discussions around it in the future.

> how to arrange wave application fold structure, or how to modulize wave app?

The framework doesn't place any constraints on how to organize code. For small apps, one file is enough. For larger apps, you are free to organize code, files, folders that best suit the app. 

There are a few small apps here as well: https://github.com/h2oai/wave-apps, which you might find useful. We plan to make more of our internal app repos public this year.

> how to write a multipage and routing to each other?

Pages in Wave are managed differently compared to typical web frameworks. Technically the Wave front-end is a single-page web application. An app has exactly one page. You can "switch" to a different page (i.e. present a different UI) by replacing all or some cards on the page with other cards. This is typically done by defining one or more [layouts](https://wave.h2o.ai/docs/layout) (which provide slots for cards), and then adding/updating/removing cards to/from those slots. For example, `q.page.drop()` clears all the cards from a page, and `del q.page['foo']` removes card `foo` from the page. This is covered here: https://wave.h2o.ai/docs/pages

> how to rewrite http headers, respond with http status?

This is related to the previous question - an app does not have direct control over the HTTP headers/status. The Wave server handles that for you. If there's something specific you want to control about the HTTP headers (security? proxying?), let us know. In typical usage, we've found that most app authors consider dealing with HTTP headers/status code an unnecessary, avoidable detail.

**Reply by vnnw** [reply]
On this page (https://wave.h2o.ai/docs/routing/#app-routing ), it says "You can host multiple apps behind a single Wave server". Care to give an example?

**Reply by lo5** [reply]
Sure.

1. Launch the Wave server (`$ ./waved`)
2. Author app `foo.py` with a `def main()` decorated with `@app('/foo')`. Launch it with `$ uvicorn foo:main`.
3. Author app `bar.py` with a `def main()` decorated with `@app('/bar')`. Launch it with `$ uvicorn bar:main`.

You should now see the two apps at http://localhost:10101/foo and http://localhost:10101/bar.

In other words, the path you pass to `@app()` determines where users can access the app.

**Comment by bwalsh** [comment]
> If there's something specific you want to control about the HTTP headers (security? proxying?), let us know.

In my scenario, there are specific headers injected by the reverse proxy that provide identity and roles.
It would be useful if https://wave.h2o.ai/docs/api/server#instance-variables-1 included

* `var headers` A [Expando](https://wave.h2o.ai/docs/api/core#h2o_wave_core_Expando) instance containing headers from the active request.

A related question: will the system populate `var access_token` if `Authorization : Bearer XXXX`  is already populated in the request? 
https://wave.h2o.ai/docs/api/server#h2o_wave_server_Auth.

**Reply by mturoci** [reply]
cc @lo5 ^^

**Reply by lo5** [reply]
@bwalsh The auth mechanism was designed to work with OAuth, so it currently ignores any inbound headers it does not care about.

We could copy all inbound headers and pass it along in the `Auth` instance - that shouldn't be a problem.

Just so that we understand this better - what's your auth/az setup high level? What other software/products are involved, and what does the auth flow look like?

**Reply by bwalsh** [reply]
Thanks very much for following up.

In a nutshell, an nginx instance will work with a rbac server to validate user and look up specific datasets the user's role has access to

<img width="660" alt="image" src="https://user-images.githubusercontent.com/47808/202358499-b069fcbb-029d-47eb-a103-cded138da344.png">

e.g.  open search's implementation

https://opensearch.org/docs/latest/security-plugin/configuration/proxy/#enable-proxy-authentication

**Reply by lo5** [reply]
Thank you for the clarification. Tracked here: https://github.com/h2oai/wave/issues/1704

**Reply by mturoci** [reply]
Hey @bwalsh need a few more details to make sure the feature is implemented correctly.

We can forward the headers from the initial `GET index.html` request. However, this `index.html` needs to download some JS/CSS resources and most importantly, register a websocket connection. I assume (and please correct me if I am wrong) that these subsequent requests also need to go through the proxy and be authenticated as well. Does the login redirect (to index.html) also set the session cookie that tells the proxy the request is properly authenticated or is there something else happening?

**Comment by bwalsh** [comment]
Thanks very much

**Reply by MaksymSofer** [reply]
code from this open PR:
https://github.com/h2oai/wave/pull/1770
works and i can manipulate over headers - very nice work guys
waiting for this to be merged

**Reply by mturoci** [reply]
Released in 0.25 for anyone who needs to access the headers - https://wave.h2o.ai/docs/security#accessing-forwarded-headers
